### PR TITLE
fixed snappy decoder bug:   `HadoopSnappyDecoder::buffer` is invalid when copy multi blocks input data at the same place

### PR DESCRIPTION
**Pull request by lgbo-ustc**
Changelog category (leave one):

- Bug Fix 

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

When the inner readbuffer's buffer size is too small, NEED_MORE_INPUT in `HadoopSnappyDecoder` will run multi times (>=3)for one compressed block. This makes the input data be copied into the wrong place in  `HadoopSnappyDecoder::buffer`.

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by vdimir** [comment]
I checked if we have tests for snappy, and found one that was disabled. Could you add commit from https://github.com/vdimir/ClickHouse/tree/snappy_01060_avro to enable it into this PR?

**Review by vdimir — APPROVED** [review]
Let's wait CI

Also I checked if we have tests for snappy, and found one that was disabled. Could you add commit from [https://github.com/vdimir/ClickHouse/tree/snappy_01060_avro](https://github.com/vdimir/ClickHouse/tree/snappy_01060_avro?rgh-link-date=2022-03-09T13%3A26%3A00Z) to enable it into this PR?

File: src/IO/tests/gtest_hadoop_snappy_decoder.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
...w0MDUwotoJDi4PCDc0N6aHARgyMjMuODMxpgYRwmcRGDIxMi4xNjWqSgIQMDkuODmuzgMYMTkuNTg0MaK7CMFFADmuZQcQMDYuMzKqXwAIOS4zrl8ADu4PBDQ0qtQUGDQ3LjAzODGmFwIYMTAuOTIwMKLDAKG0DDM3MDOiYgCqNgcORgkEMzeuGwWqXQAhqwg2MDWmSQUYMjY0LjE2N6aZFBIgFgQyM6aiCRQwNi41NTSm9AcYMjczLjczNqqSABg0NS45OTIzpugPFDIxLjc3MqZ4EBwyODYuMDkyNKZAAhg0OS4yMjQzom8GDu0LCDEwNKaTBwAzDiUIADimGQkUMzM4Ljc2qlITADcOmBUAOaYNBhwyNzAuODA4N6qrCQw3LjAwppkYwT4IMjYzrg0GDDMuOTmq/xEQMjIuODOqRgkEMjQOXxKmQA0IMzAwDggVqjwREDY1LjYxsh8aCDQuOKrCBxgyNTQuNjQ2phMUISQENzmqsAwOLgsENTWqeAIQOTEuNTiuzREANw55CQAwpp8GEDI2My44rgsRFDI0LjMxNqZuBhIrFgAxqswDGDI4OS4zMzCqXwQANoHyADCmbAMUMzI4LjM2ppsDDDY1LjKBj57+Cg5PFwQ1NaoVBmFrADaqwgccMjk5LjgxMTCqdwYQMy4wODKmZwcEMzIOqBQAMaaCBRgyMjUuMTE2qtkJADEOLw8AMKYwBBgyMzAuMTQyprwPGDMwMi4wMjemiAEOzQ4MODA0M6YaAhA1NC4yNKYkBWEMDsELqmEAFDIuNjE4N6LNBxgyODMuNTM1qqUfFDk5Ljc4NKaaGQ5UEAgyNjSuqw2usgkYNDMuMDY0MZ5rAyHkCDMzOa6sHg6+CwAwpnYGDnseCDk1MqaoAsHYDDgzNjeiLgsYMjg1LjkzMqZ1EQ67IQgyNTmmMQBB2Qg0OTamuhMUMjcxLjkzqpMWBDMyDoohADSmYgChhAg2NjimeAIQMzkxLjiqyw4IOTkuDt8bpoYBDDk0LjamaQMO4hAIOTI3qqQYFDQyLjk1M6oxAAQ4NA7GHaZKIg6YCwwxNzYzpiQXFDkwLjk0OKqqAhQ5Ny4yNzSmvwQANg54GKq/CA4AIQg1MzOm/wMUNTYuNzQ2phcCHDM0NS4wOTEyoswHDoAQCDA5M6rOGRA5MS42N6ZPGyQyNzUuNzExMTIK";
```
[the remark this answers is not in the document]
Comment (lgbo-ustc) [reply src/IO/tests/gtest_hadoop_snappy_decoder.cpp:22]: Ok

**Comment by lgbo-ustc** [comment]
1. fixed code style in [gtest_hadoop_snappy_decoder.cpp](https://github.com/ClickHouse/ClickHouse/pull/35116/files/8dc8c87fd1cea08e7e9875dd7774b02b1f3eef63#diff-32a1a18cd4cf0198a6ccaefc0727d496c89c31c786b867490fe572322410f452)
2. enable snappy test in [01060_avro.sh](https://github.com/vdimir/ClickHouse/commit/d304fa31de433af6b7df4d9f7ca3d851eec61dba#diff-f709fcd9cef80809c0d7ccda3cb5ac33638802e2198e6ed5fe41c0aba96db973)

**Pull request merged**